esp/esdllib/esdl_script.cpp

```diff
@@ -69,44 +85,89 @@ static inline StringBuffer &makeOperationTagName(StringBuffer &s, const StringBu
+    bool m_withVariables = false;
+
+public:
+    CEsdlTransformOperationWithChildren(XmlPullParser &xpp, StartTag &stag, const StringBuffer &prefix, bool withVariables, esdlOperationsFactory_t factory) : CEsdlTransformOperationBase(xpp, stag, prefix), m_withVariables(withVariables)
+    {
+        //load children
+        if (factory)
+            factory(m_children, xpp, prefix, withVariables, m_ignoreCodingErrors);
+        else
+            createEsdlTransformOperations(m_children, xpp, prefix, withVariables, m_ignoreCodingErrors);
+    }
+
+    virtual ~CEsdlTransformOperationWithChildren(){}
+
+    virtual bool processChildren(IEsdlScriptContext * scriptContext, IXpathContext * targetContext, IXpathContext * sourceContext)
     {
-        m_tagname.set(tree->queryName());
-        if (tree->hasProp("@trace"))
-            m_traceName.set(tree->queryProp("@trace"));
-        if (tree->hasProp("@_crtTarget"))
-            m_mergedTarget.set(tree->queryProp("@_crtTarget"));
-        m_ignoreCodingErrors = tree->getPropBool("@optional", false);
+        if (!m_children.length())
+            return false;
+
+        Owned<CXpathContextScope> scope = m_withVariables ? new CXpathContextScope(sourceContext, m_tagname) : nullptr;
+        bool ret = false;
+        ForEachItemIn(i, m_children)
+        {
+            if (m_children.item(i).process(scriptContext, targetContext, sourceContext))
+                ret = true;
+        }
+        return ret;
     }
 
-    virtual const char *queryMergedTarget() override
+    virtual void toDBGLog () override
     {
-        return m_mergedTarget;
+    #if defined(_DEBUG)
+        ForEachItemIn(i, m_children)
+            m_children.item(i).toDBGLog();
+    #endif
     }
 };
 
+class CEsdlTransformOperationWithoutChildren : public CEsdlTransformOperationBase
+{
+public:
+    CEsdlTransformOperationWithoutChildren(XmlPullParser &xpp, StartTag &stag, const StringBuffer &prefix) : CEsdlTransformOperationBase(xpp, stag, prefix)
+    {
+        if (xpp.skipSubTreeEx())
+            esdlOperationError(ESDL_SCRIPT_Error, m_tagname, "should not have child tags", m_traceName, !m_ignoreCodingErrors);
+    }
+
+    virtual ~CEsdlTransformOperationWithoutChildren(){}
+
+};
 
-class CEsdlTransformOperationVariable : public CEsdlTransformOperationBase
+class CEsdlTransformOperationVariable : public CEsdlTransformOperationWithChildren
 {
 protected:
     StringAttr m_name;
     Owned<ICompiledXpath> m_select;
 
 public:
-    CEsdlTransformOperationVariable(IPropertyTree *tree, const StringBuffer &prefix) : CEsdlTransformOperationBase(tree, prefix)
+    CEsdlTransformOperationVariable(XmlPullParser &xpp, StartTag &stag, const StringBuffer &prefix) : CEsdlTransformOperationWithChildren(xpp, stag, prefix, true, nullptr)
     {
         if (m_traceName.isEmpty())
-            m_traceName.set(tree->queryProp("@name"));
-        m_name.set(tree->queryProp("@name"));
+            m_traceName.set(stag.getValue("name"));
```

Comment:
Might be beneficial to have this logic in the base class to ensure that all operations default to the name as traceName when its empty.